pyigm/fN/fnmodel.py

```diff
@@ -0,0 +1,456 @@
+                if kk == 0:  # LyaF
+                    zcuts = self.param[2][2:4]
+                    gamma = self.param[2][4:]
+                else:        # DLA
+                    zcuts = [self.param[3][2]]
+                    gamma = self.param[3][3:]
+                zcuts = [0] + list(zcuts) + [999.]
+                Aval = self.param[2+kk][0]
+                # Cut on z
+                for ii in range(1,len(zcuts)):
+                    izcut = np.where( (z_val < zcuts[ii]) &
+                                      (z_val > zcuts[ii-1]) )[0]
+                    liz = len(izcut)
+                    # Evaluate (at last!)
+                    if (ii <=2) & (liz > 0):
+                        fz[izcut, kk] = Aval * ( (1+z_val[izcut]) /
+                                                 (1+zcuts[1]) )**gamma[ii-1]
+                    elif (ii == 3) & (liz > 0):
+                        fz[izcut, kk] = Aval * ( ( (1+zcuts[2]) /
+                                                   (1+zcuts[1]) )**gamma[ii-2] *
+                                                    ((1+z_val[izcut]) / (1+zcuts[2]) )**gamma[ii-1] )
+            # dX/dz
+            dXdz = pyigmu.cosm_xz(z_val, cosmo=cosmo, flg_return=1)
+
+            # Final steps
+            if flg_1D == 1: # 
+                fnX = np.sum(fz * 10.**log_gN, 1) / dXdz
+                log_fNX = np.log10(fnX)
+            else: 
+                # Generate the matrix
+                fnz = np.zeros((lenNHI, lenz))
+                for kk in range(ncomp):
+                    fnz += np.outer(10.**log_gN[:, kk], fz[:, kk])
+                # Finish up
+                log_fNX = np.log10(fnz) - np.log10( np.outer(np.ones(lenNHI), dXdz) )
+        else: 
+            raise ValueError('fN.model: Not ready for this model type {:%s}'.format(self.mtype))
+
+        # Return
+        if (lenNHI + lenz) == 2:
+            return log_fNX.flatten()[0]  # scalar
+        else:
+            return log_fNX
+
+    def mfp(self, zem, neval=5000, cosmo=None, zmin=0.6):
+        """ Calculate mean free path
+
+        Parameters
+        ----------
+        zem : float
+          Redshift of source
+        cosmo : astropy.cosmology, optional
+          Cosmological model to adopt (as needed)
+        neval : int, optional
+          Discretization parameter (5000)
+        zmin: float, optional
+          Minimum redshift in the calculation (0.5)
+
+        Returns
+        -------
+        mfp : Quantity
+          Mean free path from zem (physical Mpc)
+        """
+        # Imports
+        from pyigm.fN import tau_eff as pyteff
+        from astropy import cosmology
+
+        # Cosmology
+        if cosmo is None:
+            cosmo = cosmology.core.FlatLambdaCDM(70., 0.3)
+
+        # Calculate teff
+        zval, teff_LL = pyteff.lyman_limit(self, zmin, zem, N_eval=neval, cosmo=cosmo)
+
+        # Find tau=1
+        imn = np.argmin(np.fabs(teff_LL-1.))
+        if np.fabs(teff_LL[imn]-1.) > 0.02:
```

Comment:
Yes, but I had a reason for this once. 
And I've forgotten it. ;)

Now interpolating.